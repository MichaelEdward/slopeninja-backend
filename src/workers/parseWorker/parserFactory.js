import cheerio from 'cheerio';
import { AllHtmlEntities } from 'html-entities';

const htmlEntities = new AllHtmlEntities();

export const removeBackSlashes = (text) => {
  const decoded = text ? text.toString().replace(/\\/g, '') : null;
  return decoded;
}

export const xyz = (text) => {
  const decoded = text ? text.toString().replace(/\u0026amp;/gu, '__JAKE__') : null;
  return decoded;
}

export const decodeEntities = (text) => {
  // best way I found to normalize unicode in input
  const unicodeInput = JSON.stringify(JSON.parse(text.toString()));
  return htmlEntities.decode(unicodeInput);
}

export const createHtmlParser = (key, parser, preParser = t => t) => async (html) => {
  const $ = cheerio.load(preParser(html), { decodeEntities: true });

  const keyData = await parser($);

  return {
    [key]: keyData
  };
}

export const createJSONParser = (key, parser, preParser = t => t) => async (data = "{}") => {
  let json = JSON.parse(preParser(data));

  let d;
  if (json) {
    d = json;
  }

  const keyData = await parser(d);

  return {
    [key]: keyData
  };
}
