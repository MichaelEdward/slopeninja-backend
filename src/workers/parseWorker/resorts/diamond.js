import cheerio from 'cheerio';

import {
  degreeOrNull,
  inchOrNull,
  resortStatusOrNull,
  weatherStatusOrNull,
} from '../util';

const initialWeather = {
  status: null,
  weatherIcon: null,
  temprature: null,
  baseCondition: null,
  newSnow: null,
  snowDepthBase: null,
  snowDepthSummit: null,
};

const initialLifts = {
  total: null,
  open: null,
};

const initialTrails = {
  total: null,
  open: null,
};

export const parseDiamondSnow = async ($) => {
  const status = $('.surface-conditions p').first().text().trim();
  const weatherIcon = $('.weather-condition .weather-condition-wrapper .weather-forecast').first().text().trim();
  const temprature = $('.conditions-overlay .row.weather-row .large-4.columns').first().text().trim();
  //24 Hours
  const newSnow24Hr = $('.conditions-overlay .row.weather-row .large-4.columns .weather-data').slice(3,4).text().trim();
  //Base
  const snowDepthBase = $('.conditions-overlay .row.weather-row .large-4.columns .weather-data').slice(1,2).text().trim();
  const snowDepthSummit = $('.conditions-overlay .row.weather-row .peak-depth.large-4.columns .weather-data').first().text().trim();
  return {
    ...initialWeather,
    weatherIcon: weatherStatusOrNull(weatherIcon),
    status: resortStatusOrNull(status),
    temprature: degreeOrNull(temprature),
    newSnow: inchOrNull(newSnow24Hr),
    snowDepthBase: inchOrNull(snowDepthBase),
    snowDepthSummit: inchOrNull(snowDepthSummit),
  };
}

export const parseDiamondLifts = async ($) => {
  return {
    ...initialLifts,
  };
}

export const parseDiamondTrails = async ($) => {
  return {
    ...initialTrails,
  };
}
