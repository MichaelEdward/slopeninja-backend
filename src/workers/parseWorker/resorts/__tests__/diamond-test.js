import fs from 'fs';
import { parseDiamondSnow, parseDiamondLifts, parseDiamondTrails } from '../diamond';
import { createHtmlParser } from '../../parserFactory';

test('fetches Diamond snow data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/diamond-weather.html`);

  const resortData = await createHtmlParser('snow', parseDiamondSnow)(htmlText);
  expect(resortData).toEqual({
    snow: {
      status: 'closed',
      weatherIcon: 'clear',
      temprature: 66,
      baseCondition: null,
      newSnow: 0,
      snowDepthBase: 84,
      snowDepthSummit: 132,
    },
  });
})

test('fetches all null for nonexisting snow values', async () => {
  const resortData = await createHtmlParser('snow', parseDiamondSnow)('<html></html>');
  expect(resortData).toEqual({
    snow: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    }
  });
});

test('fetches Diamond lifts data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/diamond-weather.html`);
  const resortData = await createHtmlParser('lifts', parseDiamondLifts)(htmlText);
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches all null for nonexisting lift values', async () => {
  const resortData = await createHtmlParser('lifts', parseDiamondLifts)('<html></html>');
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches Diamond trails data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/diamond-weather.html`);
  const resortData = await createHtmlParser('trails', parseDiamondTrails)(htmlText);
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches all null for nonexisting trails values', async () => {
  const resortData = await createHtmlParser('trails', parseDiamondTrails)('<html></html>');
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});
