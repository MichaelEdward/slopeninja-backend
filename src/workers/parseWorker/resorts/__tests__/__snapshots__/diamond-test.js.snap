// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches Diamond lift list correctly 1`] = `
Object {
  "liftList": Array [
    Object {
      "category": null,
      "name": "Crystal Express Lift",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Lakeview Lift",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Lodgepole Lift",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Schoolhouse Lift",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Red Fox Lift",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Ridge Lift",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Terrain Park",
      "status": "closed",
    },
  ],
}
`;

exports[`fetches Diamond trail list correctly 1`] = `
Object {
  "trailList": Array [
    Object {
      "category": "Crystal Express Lift",
      "level": "intermediate",
      "name": "Crystal Ridge",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "intermediate",
      "name": "Golden Eagle Bowl",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "intermediate",
      "name": "Great Flume",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "intermediate",
      "name": "Lakeview",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "intermediate",
      "name": "Spillway",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "intermediate",
      "name": "Sunnyside",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "advanced",
      "name": "Battle Born",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "advanced",
      "name": "Diamondback",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "advanced",
      "name": "Lightning",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "advanced",
      "name": "Luggi's",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "advanced",
      "name": "Powder",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "advanced",
      "name": "Solitude Canyon",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "advanced",
      "name": "The Glades",
      "status": "closed",
    },
    Object {
      "category": "Crystal Express Lift",
      "level": "advanced",
      "name": "Thunder",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "intermediate",
      "name": "Cappy's Corner",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "intermediate",
      "name": "Dusty's Delight",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "beginner",
      "name": "Freeway",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "intermediate",
      "name": "Lower Showoff",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "intermediate",
      "name": "Penguin",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "intermediate",
      "name": "Popular",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "intermediate",
      "name": "Ridge Run",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "intermediate",
      "name": "Wiggle",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "advanced",
      "name": "Chute",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "advanced",
      "name": "Corkscrew",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "advanced",
      "name": "F.I.S.",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "advanced",
      "name": "G.S.",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "advanced",
      "name": "O God",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "advanced",
      "name": "Slalom Glade",
      "status": "closed",
    },
    Object {
      "category": "Lakeview Lift",
      "level": "advanced",
      "name": "Upper Showoff",
      "status": "closed",
    },
    Object {
      "category": "Lodgepole Lift",
      "level": "beginner",
      "name": "Lodge Pole",
      "status": "closed",
    },
    Object {
      "category": "Schoolhouse Lift",
      "level": "beginner",
      "name": "School Yard",
      "status": "closed",
    },
  ],
}
`;
